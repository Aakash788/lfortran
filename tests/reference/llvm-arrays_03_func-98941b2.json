{
    "basename": "llvm-arrays_03_func-98941b2",
    "cmd": "lfortran --no-color --show-llvm {infile} -o {outfile}",
    "infile": "tests/../integration_tests/arrays_03_func.f90",
    "infile_hash": "9feab893fd240821f352bd8b0c12cb132cf39c706950d08f5a41cc54",
    "outfile": null,
    "outfile_hash": null,
    "stdout": "llvm-arrays_03_func-98941b2.stdout",
    "stdout_hash": "bd2f345e61d1f8493e6f8e56fb9c7cfa3cb70ac802f0612558830b21",
    "stderr": null,
    "stderr_hash": null,
    "returncode": 0
}