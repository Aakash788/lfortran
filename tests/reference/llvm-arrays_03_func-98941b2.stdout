; ModuleID = 'LFortran'
source_filename = "LFortran"

%size_arg = type { %dimension_descriptor*, i32 }
%dimension_descriptor = type { i32, i32, i32, i32 }
%array = type { i32*, i32, %dimension_descriptor*, i1 }

@0 = private unnamed_addr constant [4 x i8] c"%d\0A\00", align 1
@1 = private unnamed_addr constant [12 x i8] c"ERROR STOP\0A\00", align 1

declare void @__module_lfortran_intrinsic_builtin_move_alloc(i32*, i32*)

define i32 @__module_lfortran_intrinsic_builtin_size(%size_arg* %x, i32* %dim, i32* %kind) {
.entry:
  %size = alloca i32, align 4
  %0 = getelementptr %size_arg, %size_arg* %x, i32 0, i32 0
  %1 = load %dimension_descriptor*, %dimension_descriptor** %0, align 8
  %2 = getelementptr %size_arg, %size_arg* %x, i32 0, i32 1
  %3 = load i32, i32* %2, align 4
  store i32 1, i32* %size, align 4
  %4 = alloca i32, align 4
  store i32 0, i32* %4, align 4
  br label %loop.head

loop.head:                                        ; preds = %loop.body, %.entry
  %5 = load i32, i32* %4, align 4
  %6 = icmp slt i32 %5, %3
  br i1 %6, label %loop.body, label %loop.end

loop.body:                                        ; preds = %loop.head
  %7 = load i32, i32* %4, align 4
  %8 = load i32, i32* %size, align 4
  %9 = getelementptr inbounds %dimension_descriptor, %dimension_descriptor* %1, i32 %7
  %10 = getelementptr %dimension_descriptor, %dimension_descriptor* %9, i32 0, i32 3
  %11 = load i32, i32* %10, align 4
  %12 = mul i32 %8, %11
  store i32 %12, i32* %size, align 4
  %13 = add i32 %7, 1
  store i32 %13, i32* %4, align 4
  br label %loop.head

loop.end:                                         ; preds = %loop.head
  br label %return

return:                                           ; preds = %loop.end
  %14 = load i32, i32* %size, align 4
  ret i32 %14
}

define i32 @mysum(%array* %a) {
.entry:
  %i = alloca i32, align 4
  %r = alloca i32, align 4
  store i32 0, i32* %r, align 4
  store i32 0, i32* %i, align 4
  br label %loop.head

loop.head:                                        ; preds = %loop.body, %.entry
  %0 = load i32, i32* %i, align 4
  %1 = add i32 %0, 1
  %2 = alloca %size_arg, align 8
  %3 = getelementptr %array, %array* %a, i32 0, i32 2
  %4 = load %dimension_descriptor*, %dimension_descriptor** %3, align 8
  %5 = getelementptr %size_arg, %size_arg* %2, i32 0, i32 0
  store %dimension_descriptor* %4, %dimension_descriptor** %5, align 8
  %6 = getelementptr %size_arg, %size_arg* %2, i32 0, i32 1
  store i32 25297008, i32* %6, align 4
  %7 = alloca i32, align 4
  %8 = alloca i32, align 4
  %9 = call i32 @__module_lfortran_intrinsic_builtin_size(%size_arg* %2, i32* %7, i32* %8)
  %10 = icmp sle i32 %1, %9
  br i1 %10, label %loop.body, label %loop.end

loop.body:                                        ; preds = %loop.head
  %11 = load i32, i32* %i, align 4
  %12 = add i32 %11, 1
  store i32 %12, i32* %i, align 4
  %13 = load i32, i32* %r, align 4
  %14 = load i32, i32* %i, align 4
  %15 = getelementptr %array, %array* %a, i32 0, i32 2
  %16 = load %dimension_descriptor*, %dimension_descriptor** %15, align 8
  %17 = getelementptr inbounds %dimension_descriptor, %dimension_descriptor* %16, i32 0
  %18 = getelementptr %dimension_descriptor, %dimension_descriptor* %17, i32 0, i32 1
  %19 = load i32, i32* %18, align 4
  %20 = sub i32 %14, %19
  %21 = mul i32 1, %20
  %22 = add i32 0, %21
  %23 = getelementptr %dimension_descriptor, %dimension_descriptor* %17, i32 0, i32 3
  %24 = load i32, i32* %23, align 4
  %25 = mul i32 1, %24
  %26 = getelementptr %array, %array* %a, i32 0, i32 0
  %27 = load i32*, i32** %26, align 8
  %28 = getelementptr inbounds i32, i32* %27, i32 %22
  %29 = load i32, i32* %28, align 4
  %30 = add i32 %13, %29
  store i32 %30, i32* %r, align 4
  br label %loop.head

loop.end:                                         ; preds = %loop.head
  br label %return

return:                                           ; preds = %loop.end
  %31 = load i32, i32* %r, align 4
  ret i32 %31
}

define i32 @main() {
.entry:
  %i = alloca i32, align 4
  %s = alloca i32, align 4
  %x = alloca %array, align 8
  %0 = getelementptr %array, %array* %x, i32 0, i32 1
  store i32 0, i32* %0, align 4
  %1 = getelementptr %array, %array* %x, i32 0, i32 2
  %2 = alloca i32, align 4
  store i32 1, i32* %2, align 4
  %3 = load i32, i32* %2, align 4
  %4 = alloca %dimension_descriptor, i32 %3, align 8
  store %dimension_descriptor* %4, %dimension_descriptor** %1, align 8
  %5 = load %dimension_descriptor*, %dimension_descriptor** %1, align 8
  %6 = getelementptr inbounds %dimension_descriptor, %dimension_descriptor* %5, i32 0
  %7 = getelementptr %dimension_descriptor, %dimension_descriptor* %6, i32 0, i32 0
  %8 = getelementptr %dimension_descriptor, %dimension_descriptor* %6, i32 0, i32 1
  %9 = getelementptr %dimension_descriptor, %dimension_descriptor* %6, i32 0, i32 2
  %10 = getelementptr %dimension_descriptor, %dimension_descriptor* %6, i32 0, i32 3
  store i32 1, i32* %7, align 4
  store i32 1, i32* %8, align 4
  store i32 10, i32* %9, align 4
  %11 = load i32, i32* %9, align 4
  %12 = load i32, i32* %8, align 4
  %13 = sub i32 %11, %12
  %14 = add i32 %13, 1
  store i32 %14, i32* %10, align 4
  %15 = alloca i32, align 4
  store i32 10, i32* %15, align 4
  %16 = getelementptr %array, %array* %x, i32 0, i32 0
  %17 = load i32, i32* %15, align 4
  %18 = alloca i32, i32 %17, align 4
  store i32* %18, i32** %16, align 8
  store i32 0, i32* %i, align 4
  br label %loop.head

loop.head:                                        ; preds = %loop.body, %.entry
  %19 = load i32, i32* %i, align 4
  %20 = add i32 %19, 1
  %21 = alloca %size_arg, align 8
  %22 = getelementptr %array, %array* %x, i32 0, i32 2
  %23 = load %dimension_descriptor*, %dimension_descriptor** %22, align 8
  %24 = getelementptr %size_arg, %size_arg* %21, i32 0, i32 0
  store %dimension_descriptor* %23, %dimension_descriptor** %24, align 8
  %25 = getelementptr %size_arg, %size_arg* %21, i32 0, i32 1
  store i32 25297008, i32* %25, align 4
  %26 = alloca i32, align 4
  %27 = alloca i32, align 4
  %28 = call i32 @__module_lfortran_intrinsic_builtin_size(%size_arg* %21, i32* %26, i32* %27)
  %29 = icmp sle i32 %20, %28
  br i1 %29, label %loop.body, label %loop.end

loop.body:                                        ; preds = %loop.head
  %30 = load i32, i32* %i, align 4
  %31 = add i32 %30, 1
  store i32 %31, i32* %i, align 4
  %32 = load i32, i32* %i, align 4
  %33 = getelementptr %array, %array* %x, i32 0, i32 2
  %34 = load %dimension_descriptor*, %dimension_descriptor** %33, align 8
  %35 = getelementptr inbounds %dimension_descriptor, %dimension_descriptor* %34, i32 0
  %36 = getelementptr %dimension_descriptor, %dimension_descriptor* %35, i32 0, i32 1
  %37 = load i32, i32* %36, align 4
  %38 = sub i32 %32, %37
  %39 = mul i32 1, %38
  %40 = add i32 0, %39
  %41 = getelementptr %dimension_descriptor, %dimension_descriptor* %35, i32 0, i32 3
  %42 = load i32, i32* %41, align 4
  %43 = mul i32 1, %42
  %44 = getelementptr %array, %array* %x, i32 0, i32 0
  %45 = load i32*, i32** %44, align 8
  %46 = getelementptr inbounds i32, i32* %45, i32 %40
  %47 = load i32, i32* %i, align 4
  store i32 %47, i32* %46, align 4
  br label %loop.head

loop.end:                                         ; preds = %loop.head
  %48 = alloca %array, align 8
  %49 = call i32 @mysum(%array* %x)
  store i32 %49, i32* %s, align 4
  %50 = load i32, i32* %s, align 4
  call void (i8*, ...) @_lfortran_printf(i8* getelementptr inbounds ([4 x i8], [4 x i8]* @0, i32 0, i32 0), i32 %50)
  %51 = load i32, i32* %s, align 4
  %52 = icmp ne i32 %51, 55
  br i1 %52, label %then, label %else

then:                                             ; preds = %loop.end
  call void (i8*, ...) @_lfortran_printf(i8* getelementptr inbounds ([12 x i8], [12 x i8]* @1, i32 0, i32 0))
  call void @exit(i32 1)
  br label %ifcont

else:                                             ; preds = %loop.end
  br label %ifcont

ifcont:                                           ; preds = %else, %then
  ret i32 0
}

declare void @_lfortran_printf(i8*, ...)

declare void @exit(i32)
